### DISCUSSION TITLE
Demo of Rust and Axum -- seeking feedback

### DISCUSSION THREAD

**Discussion by joelparkerhenderson** [tokio-rs/axum#847] [General]
Hi, Axum looks great! I'm writing a demo of Rust and Axum that's a tutorial especially for new users. I'm seeking feedback from Axum users and developers, for ideas, suggestions, any code quality feedback, etc. Thank you. Best, Joel. 

https://github.com/joelparkerhenderson/demo-rust-axum

**Comment by davidpdrsn** [comment]
Awesome! Feel free to make a PR adding it to ECOSYSTEM.md when you feel its ready 😊

Some quick things I noticed:
- We generally write axum all lowercase, though not totally consistently. I prefer it lowercase though 😅
- Don't use `*` when declaring dependencies. Use version numbers instead, otherwise you dunno which version people will get in the future.
- You don't need to write `extern crate _` anymore. Rust does that automatically.
- Should probably explain the syntax `Json(payload): Json<serde_json::Value>` in function arguments. It confuses some beginners.
- You can also initialize the books inside the call to `Lazy::new`. Doesn't _need_ to be a separate function.

**Reply by joelparkerhenderson** [reply]
Great feedback, thank you for the close read. I'll make the changes and do the PR this week. Much appreciated!

**Reply by joelparkerhenderson** [reply]
PR ready: https://github.com/tokio-rs/axum/pull/852

I fielded all your comments. I added a few more similar handler examples, improved the documentation by using rustdoc, refactored to improve clarity, etc.

By the way, the ECOSYSTEM file is a superb idea-- more projects should feature these IMHO. I'm reading about Embark Studios this week, and you all are doing excellent work with open source.

**Reply by jplatte** [reply]
It looks like to follow your example, one would rerun the app a lot. Given graceful shutdown isn't set up, I think in some circumstances the app might hang when you want to abort it. See the graceful shutdown example.

Also in 10., you set a request content-type of json even though you don't send any json. I would recommend you remove that and add the `-I` flag to the curl command so it shows the response headers.

**Reply by joelparkerhenderson** [reply]
^ Good feedback @jplatte, thank you. I'll add those next.

**Reply by joelparkerhenderson** [reply]
Added. See new README section 2. I also added info from the hyper docs and the tokio docs, which both have good explanations IMHO. 

Does this update add enough for what you're suggesting?

In particular, the demo code now handles the tokio signal CTRL-C. I see your example that handles more, such as Unix and non-Unix areas. For this tutorial, how important are these to you? 

My goal is to get the reader serving content quickly because I believe that this gets a beginner to reach an "aha" moment.

**Reply by jplatte** [reply]
Yeah it makes a lot of sense to me to only do the Ctrl+C thing. People will realize that they generally have to do that stuff themselves and can choose a different way of triggering graceful shutdow.

I just noticed a new thing though, you created a `Response` manually for the PNG example:

```rust
let body = axum::body::Full::from(::base64::decode(png).unwrap());
let mut response = axum::response::Response::new(body);
response.headers_mut().insert(
    ::http::header::CONTENT_TYPE,
    ::http::header::HeaderValue::from_static("image/png")
);
response
```

This can greatly be simplified to

```rust
(
    axum::response::Headers([(http::header::CONTENT_TYPE, "image/png")]),
    base64::decode(png).unwrap(),
)
```

**Reply by joelparkerhenderson** [reply]
Much simpler, thank you. I updated it now.
